### ISSUE TITLE
emar duplicates object files when used in incremental builds

### ISSUE THREAD

**Issue by r0mai** [emscripten-core/emscripten#9653]
Sorry, this became a long issue, describing a relatively simple bug.
TL;DR: incremental build with cmake and ninja is broken, because emar duplicates object files in static library archives.

---

Consider a very basic build environment using the official emscripten cmake toolchain and [ninja](https://ninja-build.org/) as the build system and emscripten 1.38.47-upstream.

`foo.cpp`:

```cpp
int foo() {
    return 1;
}
```

`main.cpp`:

```cpp
#include <cstdio>

int foo();

int main() {
    printf("%d\n", foo());
}
```

setup steps (set EMSCRIPTEN_ROOT appropriately):
```sh
mkdir -p ./build
cd ./build
cmake -DCMAKE_TOOLCHAIN_FILE=${EMSCRIPTEN_ROOT}/cmake/Modules/Platform/Emscripten.cmake -G "Ninja" ..
```

build:
```sh
$ ninja
```

run:
```
$ node exe.js
1
```

now modify `foo.cpp`, so that it returns 2:
```cpp
int foo() {
    return 2;
}
```

build again:
```sh
$ ninja
```

run again:
```sh
$ node exe.js
1
```

It still prints `1`.

The underlying issue is easier to understand, if we break down what commands are generated by cmake and then executed by ninja (without unnecessary flags):

First build builds objects files from `foo.cpp` and `main.cpp`, creates an archive `libfoo.a` from `foo.cpp.o` and then links `exe.js` together.
```sh
em++ -c foo.cpp -o foo.cpp.o
emar rc libfoo.a foo.cpp.o
em++ -c main.cpp -o main.cpp.o
em++ -o exe.js main.cpp.o libfoo.a
```

So far so good, this still prints the right thing. Also note, that libfoo.a only contains a single object file:

```sh
$ emar -t libfoo.a
foo.cpp.o
```

Second, incremental build only rebuilds foo.cpp (as it was modified), re-creates `libfoo.a`, and re-links `exe.js`:

```sh
em++ -c foo.cpp -o foo.cpp.o
emar rc libfoo.a foo.cpp.o
em++ -o exe.js main.cpp.o libfoo.a
```

After this, we can see, that `libfoo.a` contains two object files. The original one, and one with a hash at the end:
```sh
$ emar -t libfoo.a
foo.cpp.o
foo_48f75e22.cpp.o
```

The first object file is still the same object file from the previous build with the original `int foo()` function inside, which takes priority during linking.

This is clearly a bug at this point. Old object files, from previous builds should not appear in archives.

`emar` is a wrapper around `llvm-ar`, [which adds this extra hash](https://github.com/emscripten-core/emscripten/blob/1.38.47/emar.py#L73) postfix to files which already exist in the output archive. 

The hashing was introduced in reponse to handling files with the same basename (which bound to happen in large codebases): https://github.com/emscripten-core/emscripten/issues/2142

I also confirmed, that always hashing fixes the issue. This is the crude diff I used: https://github.com/emscripten-core/emscripten/compare/incoming...r0mai:emar-always-hash

Also note, that `"Unix Makefiles"` cmake generator is not affected as it uses response files, which are not touched by emar.

**Comment by sbc100** [comment]
Thanks for the report @r0mai.  Sounds like your patch makes sense to land.  Can you open a PR.

BTW, this issue is fastcomp-only and we are planning to remove support for fastcomp pretty soon.   Is there anything stopping you from switching to the llvm backend?

**Comment by r0mai** [comment]
We're actually in the process of upgrading to the llvm backend, and I confirmed that this is reproducible with both upstream and fastcomp backends:
```
emsdk activate 1.38.47-upstream
emsdk activate 1.38.47
```

**Comment by sbc100** [comment]
Ok thats good to know.  Would you mind opening a PR with your fix?

**Comment by r0mai** [comment]
of course: https://github.com/emscripten-core/emscripten/pull/9654
